compact.go

```diff
@@ -835,12 +846,54 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 			return errors.Wrap(err, "write label index")
 		}
 	}
+	names = append(names, apkName)
+	sort.Strings(names)
+
+	if idxw, ok := indexw.(*index.Writer); ok {
+		idxw.HintPostingsWriteCount(numLabelValues)
+	}
 
-	for _, l := range postings.SortedKeys() {
-		if err := indexw.WritePostings(l.Name, l.Value, postings.Get(l.Name, l.Value)); err != nil {
-			return errors.Wrap(err, "write postings")
+	postingBuf := make([]uint64, 0, 1000000)
+	seriesMap := set.SeriesMap()
+	var bigEndianPos index.Postings = index.NewBigEndianPostings(nil)
+	var listPos = index.NewListPostings(nil).(*index.ListPostings)
+	for _, n := range names {
+		s = s[:0]
+		if n == apkName {
+			s = append(s, apkValue)
+		} else {
+			for v := range values[n] {
+				s = append(s, v)
+			}
+			sort.Strings(s)
+		}
+
+		for _, v := range s {
+			postingBuf = postingBuf[:0]
+			for i, ir := range indexReaders {
+				bigEndianPos, err = ir.Postings(n, v, bigEndianPos)
+				if err != nil {
+					return errors.Wrap(err, "read postings")
+				}
+				for bigEndianPos.Next() {
+					if newVal, ok := seriesMap[i][bigEndianPos.At()]; ok {
+						// idx is the index at which newVal exists or index at which we need to insert.
```

Comment:
Insertion sort is n^2. How about putting it all into a map, and then sorting the keys? That'd be nlogn.

Reply:
It was like that before. But that seemed to cause a lot of allocs because I cannot re-use the map. With improvements in other areas for allocs, I will try it again.

Reply:
Now I remember why I removed - it was a significant part of memory usage. I tried it again and memory savings become half. Also for some reason time also goes up (both reusing maps by deleting keys and not reusing map).

Reply:
Maybe the reduced allocs/memory covers up the use of maps and heap/quick sort.